### PR TITLE
Fix cuSTL/reduce

### PR DESCRIPTION
**Pull request by Heikman**
fix cuSTL/reduce by wrapping PMacc/reduce

 - cuSTL/reduce put out completely wrong values.

### CODE REVIEW

**Comment by Heikman** [comment]
please wait for #935 to be merged.

**Comment by psychocoderHPC** [comment]
#935 is merged, please rebase against `dev`

File: src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.tpp [hunk 1]
```diff
@@ -36,124 +37,74 @@ namespace algorithm
...
+    
+/**
+ * @class MappedNavigator
+ * @author Heiko Burau 
+ * @date 30/06/15
+ * @file Reduce.tpp
+ * @brief Wraps a ndim-cursor into a 1D-cursor
+ */
+template<int T_dim>
+class MappedNavigator
```
Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.tpp]: Please rename this file to `MappedNavigator.tpp`. We name files like the class inside.

Comment (Heikman) [inline src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.tpp]: `MappedNavigator` is just a helper class for the actual class `Reduce`. Therefore it's in the namespace `detail`.

Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.tpp]: Sry might be it was not well explained.
I mean, please move the implementation of `MappedNavigator` to the other accessor implementations `cuSTL/cursor/accessor`

File: src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.tpp [hunk 2]
```diff
@@ -36,124 +37,74 @@ namespace algorithm
...
+    
 namespace detail
 {
-
-template<typename BlockDim, int dim>
-struct ReduceKernel
+    
+/**
+ * @class MappedNavigator
+ * @author Heiko Burau
```
Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.tpp]: Please remove this:
@author, @date and @file

**Comment by psychocoderHPC** [comment]
Please split this pull request in feature and bug fix
  - MappedNavigator is a new feature and can merged in `dev` as separate pull request
  - add bug fix for reduce
  - for `ThermalTestSimulation.hpp` it is not clear if this is a bug fix and if it corresponds to the reduce bug

**Comment by Heikman** [comment]
`ThermalTestSimulation.hpp` is part of the bug fix. The bug fix involves a change in the interface of `Reduce`, so each reference to it has to be adapted.

**Comment by Heikman** [comment]
Please wait for #940 to be merged first.

**Comment by psychocoderHPC** [comment]
Thanks for splitting this pull request :+1: . I will check it as soon as I am back in office.

File: src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.hpp [hunk 3]
```diff
@@ -42,23 +42,17 @@ namespace kernel
...
+/* \param srcCursor Cursor located at the origin of the area of reduce
  * \param p_zone Zone of cells spanning the area of reduce
- * \param srcCursor Cursor located at the origin of the area of reduce
  * \param functor Functor with two arguments which returns the result of the reduce operation.
  *        Can also be a lambda expression.
  */
-template<typename DestCursor, typename Zone, typename SrcCursor, typename Functor>
-void operator()(const DestCursor& destCursor, const Zone& p_zone, const SrcCursor& srcCursor, const Functor& functor);
+template<typename SrcCursor, typename Zone, typename Functor>
+typename SrcCursor::ValueType operator()(const SrcCursor& srcCursor, const Zone& p_zone, const Functor& functor);
```
Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/algorithm/kernel/Reduce.hpp]: Is it possible to move the implementation of this method from `*.tpp` to `*.hpp`?

**Comment by psychocoderHPC** [comment]
#940 is merged, please rebase this branch.

**Comment by Heikman** [comment]
done

**Pull request merged**